src/ocrd/processor/base.py

```diff
@@ -188,20 +193,122 @@ def verify(self):
+            try:
+                # FIXME: add page parallelization by running multiprocessing.Pool (#322)
+                for input_file_tuple in self.zip_input_files(on_error='abort'):
+                    # FIXME: add error handling by catching exceptions in various ways (#579)
+                    # for example:
+                    # - ResourceNotFoundError → use ResourceManager to download (once), then retry
+                    # - transient (I/O or OOM) error → maybe sleep, retry
+                    # - persistent (data) error → skip / dummy / raise
+                    input_files = [None] * len(input_file_tuple)
+                    for i, input_file in enumerate(input_file_tuple):
+                        if i == 0:
+                            log.info("processing page %s", input_file.pageId)
+                        elif input_file is None:
+                            # file/page not found in this file grp
+                            continue
+                        input_files[i] = input_file
+                        if not self.download:
+                            continue
+                        try:
+                            input_files[i] = self.workspace.download_file(input_file)
+                        except ValueError as e:
+                            log.error(repr(e))
+                            log.warning("skipping file %s for page %s", input_file, input_file.pageId)
+                    self.process_page_file(*input_files)
+            except NotImplementedError:
+                # fall back to deprecated method
+                self.process()
 
-    def add_metadata(self, pcgts):
+    def process_page_file(self, *input_files) -> None:
+        """
+        Process the given ``input_files`` of the :py:attr:`workspace`,
+        representing one physical page (passed as one opened
+        :py:class:`~ocrd_models.OcrdFile` per input fileGrp)
+        under the given :py:attr:`parameter`, and make sure the
+        results get added accordingly.
+
+        (This uses process_page_pcgts, but can be overridden by subclasses
+        to handle cases like multiple fileGrps, non-PAGE input etc.)
+        """
+        log = getLogger('ocrd.processor.base')
+        input_pcgts = [None] * len(input_files)
+        for i, input_file in enumerate(input_files):
+            # FIXME: what about non-PAGE input like image or JSON ???
+            log.debug("parsing file %s for page %s", input_file.ID, input_file.pageId)
+            try:
+                input_pcgts[i] = page_from_file(input_file)
+            except ValueError as e:
+                log.info("non-PAGE input for page %s: %s", input_file.pageId, e)
+        output_pcgts = self.process_page_pcgts(*input_pcgts)
+        output_file_id = make_file_id(input_files[0], self.output_file_grp)
+        output_pcgts.set_pcGtsId(output_file_id)
+        self.add_metadata(output_pcgts)
+        # FIXME: what about save_image_file in process_page ???
```

Comment:
Sounds like a good idea to abstract that step.

Reply:
The problem is that for processors which do add derived images, it will not suffice to use `process_page_file` from the superclass and implement `process_page_pcgts`, because the latter does not yield a reference to the page ID or file ID to use with `Workspace.save_image_file`. So all those processors would have to write their own version of `process_page_file` from scratch, which I really dislike.

So perhaps it would be better to extend `process_page_pcgts` with kwargs `output_file_id` and `page_id`?

This might also be a good chance of finally fixing #343 ...

Reply:
> So all those processors would have to write their own version of process_page_file from scratch, which I really dislike.

+1

> So perhaps it would be better to extend process_page_pcgts with kwargs output_file_id and page_id?

I still do not entirely understand when the `process_page_pcgts` is required and when not. For the dummy processor, just the first element of the input is returned. If passing extra kwargs would remove the requirement for processors to reimplement the `process_page_file` from scratch - then good.

Reply:
Both process_page_file and process_page_pcgts process 1 physical page but take multi-valued (variadic) input, because the actual processor might (need to) be called with multiple input fileGrps. The difference is that process_page_file takes an OcrdFile and returns None (after saving output to the METS), while process_page_pcgts takes OcrdPage and returns OcrdPage. The latter **must** be overriden, the former **can**.

The case with ocrd-dummy is special, because with `copy_files=False` it suffices to override process_page_pcgts in the most trivial fashion, otherwise one needs to override process_page_file as well.

Reply:
Another possibility would be that `process_page_pcgts` could optionally return a `PIL.Image` besides the OcrdPage. Then our `process_page_file` could use that for `workspace.save_image_file`.

So far, we differentiated the file IDs of derived images by appending something like `.IMG-CROP`, `.IMG-DESKEW`, `.IMG-BIN` etc. That would have to become just `.IMG` here (or one expects a tuple of image and suffix...)

Reply:
> Another possibility would be that `process_page_pcgts` could optionally return a `PIL.Image` besides the OcrdPage. Then our `process_page_file` could use that for `workspace.save_image_file`.
> 
> So far, we differentiated the file IDs of derived images by appending something like `.IMG-CROP`, `.IMG-DESKEW`, `.IMG-BIN` etc. That would have to become just `.IMG` here (or one expects a tuple of image and suffix...)

In the most general case, we do in fact need more information than just the image itself: there can be multiple images (think line-level dewarping), and each needs their own clash-free suffix (for which we usually use the segment ID).

Also, the implementor of `process_page_pcgts` will need to know the final path name of the annotated image file in advance, so they can reference it in the OcrdPage. But for that path name, one in turn needs to know the file ID of the basic output OcrdPage.

So I think we should combine both ideas – add kwargs `output_file_id` and `page_id` but also specify the following return type:
- either just `OcrdPage`
- or a list/tuple of
  - at pos 0: the `OcrdPage`
  - at each following pos: a list/tuple of
    1. the `PIL.Image` to add
    2. its full `file_id`
    3. its `local_filename`

So for [ocrd-cis-ocropy-dewarp](https://github.com/bertsky/ocrd_cis/blob/38ce45bf016546b748cce65031cad3fe24a35c0d/ocrd_cis/ocropy/dewarp.py#L167-L172) this would look like:
```python
    def process_page_pcgts(self, pcgts, output_file_id=None, page_id=None):
        page = pcgts.get_Page()
        image, coords, info = self.workspace.image_from_page(page, page_id)
        ...
        ret = [pcgts]
        for region in regions:
            ...
            for line in lines:
                ...
                file_id = output_file_id + '_' + region.id + '_' + line.id + '.IMG-DEWARP'
                file_path = os.path.join(self.output_file_grp, file_id + '.png')
                ret.append((dew_image, file_id, file_path))
                line.add_AlternativeImage(AlternativeImageType(
                  filename=file_path, comments=coords['features'] + ',dewarped'))
        return ret
```

What do you think?

Reply:
I implemented that option in e2cbcb9 now.

Reply:
> This might also be a good chance of finally fixing #343 ...

I addressed that in 20a6a1c and 679ad85 now.